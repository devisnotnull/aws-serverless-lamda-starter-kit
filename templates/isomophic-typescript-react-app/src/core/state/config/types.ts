import { IResponseAction } from '../types';

export const TOGGLE_FEATURE = '@config/TOGGLE_FEATURE';

// You can use the provided IResponseAction or IRequestAction interface.
// The first Type argument is the action
// The second Type argument is the payload or request
export type ToggleFeatureAction = IResponseAction<
    typeof TOGGLE_FEATURE,
    { feature: string; toggle: boolean }
>;

export type ActionTypes = ToggleFeatureAction;
