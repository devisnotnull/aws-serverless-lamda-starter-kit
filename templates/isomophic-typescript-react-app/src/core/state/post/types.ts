import { IPostArrayResponsePayload, IPostSingularResponsePayload } from '@core/models/post';

import { IRequestAction, IGraphqlResponseAction, IGraphqlQuery } from '../types';

// Graphql input interfaces
export interface IPageQueryOptionsInput {
    options: {
        paginate: {
            page: number;
            limit: number;
        };
    };
}

export interface ICreatePostInput {
    input: {
        id: string;
        title: string;
        body: string;
    };
}

// Declare our actions

export const FETCH_ALL = '@post/FETCH_ALL';
export const FETCH_ALL_SUCCESS = '@post/FETCH_ALL_SUCCESS';
export const FETCH_ALL_ERROR = '@post/FETCH_ALL_ERROR';
export const FETCH_ALL_ABORT = '@post/FETCH_ALL_ABORT';

export const FETCH_BY_ID = '@post/FETCH_BY_ID';
export const FETCH_BY_ID_SUCCESS = '@post/FETCH_BY_ID_SUCCESS';
export const FETCH_BY_ID_ERROR = '@post/FETCH_BY_ID_ERROR';
export const FETCH_BY_ID_ABORT = '@post/FETCH_BY_ID_ABORT';

export const CREATE_POST = '@post/CREATE_POST';
export const CREATE_POST_SUCCESS = '@post/CREATE_POST_SUCCESS';
export const CREATE_POST_ERROR = '@post/CREATE_POST_ERROR';
export const CREATE_POST_ABORT = '@post/CREATE_POST_ABORT';

// You can use the provided IAction or IRequestAction interface.
// The first Type argument is the action
// The second Type argument is the payload or request
export type FetchAllAction = IRequestAction<
    typeof FETCH_ALL,
    IGraphqlQuery<IPageQueryOptionsInput>
>;
export type FetchAllSucessAction = IGraphqlResponseAction<
    typeof FETCH_ALL_SUCCESS,
    IPostArrayResponsePayload
>;
export type FetchAllErrorAction = IGraphqlResponseAction<typeof FETCH_ALL_ERROR, any>;

export type FetchAllActionTypes = FetchAllAction | FetchAllSucessAction | FetchAllErrorAction;



export type FetchByIdAction = IRequestAction<typeof FETCH_BY_ID, IGraphqlQuery<{ id: number }>>;
export type FetchByIdSucessAction = IGraphqlResponseAction<
    typeof FETCH_BY_ID_SUCCESS,
    IPostSingularResponsePayload
>;
export type FetchByIdErrorAction = IGraphqlResponseAction<typeof FETCH_BY_ID_ERROR, any>;
export type FetchByIdActionTypes = FetchByIdAction | FetchByIdSucessAction | FetchByIdErrorAction;

export type CreateAction = IRequestAction<typeof CREATE_POST, IGraphqlQuery<ICreatePostInput>>;
export type CreateActionSucessPayload = IGraphqlResponseAction<typeof CREATE_POST_SUCCESS, any>;
export type CreateActionErrorPayload = IGraphqlResponseAction<typeof CREATE_POST_ERROR, any>;
export type CreateActionTypes = CreateAction | CreateActionSucessPayload | CreateActionErrorPayload;

export type ActionTypes = FetchAllActionTypes | FetchByIdActionTypes | CreateActionTypes;
