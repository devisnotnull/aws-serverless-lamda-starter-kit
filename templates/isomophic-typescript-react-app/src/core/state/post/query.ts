import { gql } from 'redux-saga-requests-graphql';

// Graphql input interfaces
export interface IPageQueryOptionsInput {
    options: {
        paginate: {
            page: number;
            limit: number;
        };
    };
}

export interface ICreatePostInput {
    input: {
        id: string;
        title: string;
        body: string;
    };
}

// GraphQL queries
export const fetchPostsQuery = gql`
    query($options: PageQueryOptions!) {
        posts(options: $options) {
            data {
                id
                title
            }
            meta {
                totalCount
            }
        }
    }
`;

export const fetchPostsByIdQuery = gql`
    query($id: ID!) {
        post(id: $id) {
            id
            title
            body
        }
    }
`;

export const createPost = gql`
    mutation($input: CreatePostInput!) {
        createPost(input: $input) {
            id
            title
            body
        }
    }
`;

export const deletePost = gql`
    mutation($id: ID!) {
        deletePost(id: $id)
    }
`;
