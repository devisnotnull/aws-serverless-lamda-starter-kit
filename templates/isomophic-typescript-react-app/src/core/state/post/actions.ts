import { fetchPostsByIdQuery, fetchPostsQuery, createPost } from './query';

import { ActionTypes, FETCH_ALL_START, FETCH_BY_ID_START, CREATE_POST, FETCH_ALL_START_SUCCESS, FETCH_BY_ID_START_SUCCESS, CREATE_POST_SUCCESS } from './types';

import { DEFAULT_PAGE, DEFAULT_PER_PAGE } from './const';
import { IPostArrayResponsePayload, IPostSingularResponsePayload } from '@core/models/post';

//

export const fetchAllPostsRequest = (
    page = DEFAULT_PAGE,
    limit = DEFAULT_PER_PAGE
): ActionTypes => ({
    type: FETCH_ALL_START,
    request: {
        query: fetchPostsQuery,
        variables: {
            options: {
                paginate: { page, limit },
            },
        },
    },
});

export const fetchAllPostsSucess = (data: IPostArrayResponsePayload): ActionTypes => ({
    type: FETCH_ALL_START_SUCCESS,
    data
});

//

export const fetchByIdPostsRequest = (id: number): ActionTypes => ({
    type: FETCH_BY_ID_START,
    request: {
        query: fetchPostsByIdQuery,
        variables: { id },
    },
});

export const fetchByIdPostsSuccess = (data: IPostSingularResponsePayload): ActionTypes => ({
    type: FETCH_BY_ID_START_SUCCESS,
    data
});

//

export const createPostsRequest = (postInput: any): ActionTypes => ({
    type: CREATE_POST,
    request: {
        query: createPost,
        variables: {
            input: postInput,
        },
    },
});

export const createPostsRequestSuccess = (data: IPostSingularResponsePayload): ActionTypes => ({
    type: CREATE_POST_SUCCESS,
    data
});
