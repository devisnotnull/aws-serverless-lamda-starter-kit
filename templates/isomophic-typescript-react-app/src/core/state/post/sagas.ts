import { put, all, takeEvery, call } from 'redux-saga/effects';

import { IPostSingularResponsePayload, IPostArrayResponsePayload } from '@core/models/post';

import { fetchAllPostsRequest, fetchAllPostsSucess, fetchByIdPostsSuccess } from './actions';

import { IPayload, GraphQlGeneratorReturnType } from '../types';
import { fetch } from '../utils';
import { FETCH_ALL, FetchAllAction, FetchByIdAction } from './types';

// Saga calls to apollo

export function* fetchAllStart({
    request: { query, variables },
}: FetchAllAction): GraphQlGeneratorReturnType<IPostArrayResponsePayload> {
    try {
        const data: IPayload<IPostArrayResponsePayload> = yield call(fetch, query, variables);
        yield put(fetchAllPostsSucess(data.data));
    } catch (ex) {
        console.log('[Critical Error]: Unable to perform network action with error ', ex);
    }
}

export function* fetchByIdStart({
    request: { query, variables },
}: FetchByIdAction): GraphQlGeneratorReturnType<IPostSingularResponsePayload> {
    try {
        const data: IPayload<IPostSingularResponsePayload> = yield call(fetch, query, variables);
        yield put(fetchByIdPostsSuccess(data.data));
    } catch (ex) {
        console.log('[Critical Error]: Unable to perform network action with error ', ex);
    }
}

// Functional proxy calls

export function* fetchAllSaga(page?: number, limit?: number) {
    yield fetchAllStart(fetchAllPostsRequest() as FetchAllAction);
}

// Listener generator funtions

export function* postSagas() {
    yield all([fetchAllSaga(), takeEvery(FETCH_ALL, fetchAllSaga)]);
}
