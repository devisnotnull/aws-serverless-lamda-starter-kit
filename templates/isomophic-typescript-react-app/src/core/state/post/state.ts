import { IPost } from '@core/models/post';

export interface IPostState {
    readonly loading: boolean;
    readonly errors: string | undefined;
    readonly total?: number;
    readonly skip?: number;
    readonly post: IPost | undefined;
    readonly posts: IPost[];
}

export const initialState: IPostState = {
    total: 0,
    skip: 0,
    post: undefined,
    posts: [],
    errors: undefined,
    loading: false,
};
