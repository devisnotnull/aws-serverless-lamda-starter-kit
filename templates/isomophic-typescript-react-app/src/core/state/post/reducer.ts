import { Reducer } from 'redux';
import { UNABLE_TO_LOAD_ERROR } from '../consts';

import {
    ActionTypes,
    FETCH_ALL_START,
    FETCH_ALL_START_SUCCESS,
    FETCH_ALL_START_ERROR,
    FETCH_BY_ID_START,
    FETCH_BY_ID_START_SUCCESS,
    FETCH_BY_ID_START_ERROR,
} from './types';

import { IPostState, initialState } from './state';

export const postReducers: Reducer<IPostState> = (
    state: IPostState = initialState,
    action: ActionTypes
): IPostState => {
    switch (action.type) {
        //
        // Base action: FETCH_ALL_START
        //
        case FETCH_ALL_START: {
            return { ...state, loading: true, errors: undefined };
        }
        case FETCH_ALL_START_SUCCESS: {
            const payload = action.data?.posts ?? undefined;
            return {
                ...state,
                loading: false,
                errors: payload ? undefined : UNABLE_TO_LOAD_ERROR,
                total: payload?.meta?.totalCount ?? 0,
                posts: payload?.data ?? [],
            };
        }
        case FETCH_ALL_START_ERROR: {
            return { ...state, loading: false, errors: UNABLE_TO_LOAD_ERROR };
        }
        //
        // Base action: FETCH_BY_ID
        //
        case FETCH_BY_ID_START: {
            return { ...state, loading: true, errors: undefined };
        }
        case FETCH_BY_ID_START_SUCCESS: {
            const payload = action.data?.post ?? undefined;
            return {
                ...state,
                loading: false,
                errors: payload ? undefined : UNABLE_TO_LOAD_ERROR,
                ...payload,
            };
        }
        case FETCH_BY_ID_START_ERROR: {
            return { ...state, loading: false, errors: UNABLE_TO_LOAD_ERROR };
        }
        //
        //
        //
        default: {
            return state;
        }
    }
};

export default postReducers;
