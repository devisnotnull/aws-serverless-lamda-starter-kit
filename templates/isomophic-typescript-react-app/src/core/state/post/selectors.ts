import { prop, pipe, propOr } from 'ramda'

import { IPostState } from './state'
import { IState } from '../state'

export const getPosts = (state: IState): IPostState => state?.post ?? {}

export const getPostsItems = pipe(getPosts, prop('data'))

export const getPostsLoading = pipe(getPosts, prop('loading'))

export const getPostsItemsErrors = pipe(getPosts, propOr(undefined, 'errors'))
