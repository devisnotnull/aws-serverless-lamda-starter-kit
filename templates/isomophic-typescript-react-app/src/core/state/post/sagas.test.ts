import { testSaga, expectSaga } from 'redux-saga-test-plan';

import {
    fetchAllPostsRequest,
    fetchByIdPostsRequest,
    fetchAllPostsSucess,
    fetchByIdPostsSuccess,
} from './actions';
import { fetchAllSaga, fetchById, fetchAllStart, fetchByIdStart } from './sagas';
import { fetch, IPayload } from '../utils';
import { FetchAllAction, FetchByIdAction } from './types';
import { IPostArrayResponsePayload, IPostSingularResponsePayload } from '@core/models/post';
import { call } from 'redux-saga/effects';

describe('post saga: fetchAllStart', () => {
    it('saga: fetchAllStart with no params', async () => {
        const action: FetchAllAction = fetchAllPostsRequest() as FetchAllAction;
        const payload: IPayload<IPostArrayResponsePayload> = {
            data: {
                posts: {
                    data: [],
                    meta: {},
                },
            },
        };
        return expectSaga(fetchAllStart, action)
            .provide([[call(fetch, action.request.query, action.request.variables), payload]])
            .put(fetchAllPostsSucess(payload.data))
            .run();
    });

    it('saga: snapshot: fetchAllSaga with no page and limit params', async () => {
        const action: FetchAllAction = fetchAllPostsRequest() as FetchAllAction;
        const payload = (await expectSaga(fetchAllStart, action).run()).toJSON();
        expect(payload).toMatchSnapshot();
    });
});

describe('post saga: fetchByIdStart', () => {
    it('saga: fetchByIdStart with id', async () => {
        const action: FetchByIdAction = fetchByIdPostsRequest(1) as FetchByIdAction;
        const payload: IPayload<IPostSingularResponsePayload> = {
            data: {
                post: {
                    id: '',
                    title: '',
                    body: '',
                },
            },
        };
        return expectSaga(fetchByIdStart, action)
            .provide([[call(fetch, action.request.query, action.request.variables), payload]])
            .put(fetchByIdPostsSuccess(payload.data))
            .run();
    });

    it('saga: snapshot: fetchAllSaga with no page and limit params', async () => {
        const action: FetchAllAction = fetchAllPostsRequest() as FetchAllAction;
        const payload = (await expectSaga(fetchAllStart, action).run()).toJSON();
        expect(payload).toMatchSnapshot();
    });
});

describe('post saga: fetchAll', () => {
    it('saga: fetchAllSaga with no params', () => {
        testSaga(fetchAllSaga)
            //
            .next()
            //
            .put(fetchAllPostsRequest())
            //
            .next()
            //
            .isDone();
    });

    it('saga: fetchAllSaga with correct page and limit params', () => {
        testSaga(fetchAllSaga, 1, 2)
            //
            .next()
            //
            .put(fetchAllPostsRequest(1, 2))
            //
            .next()
            //
            .isDone();
    });

    it('saga: snapshot: fetchAllSaga with no page and limit params', async () => {
        const payload = (await expectSaga(fetchAllSaga, 1, 2).run()).toJSON();
        expect(payload).toMatchSnapshot();
    });
});

describe('post saga: fetchById', () => {
    it('saga: fetchById with correct input param', () => {
        testSaga(fetchById, 1)
            //
            .next()
            //
            .put(fetchByIdPostsRequest(1))
            //
            .next()
            //
            .isDone();
    });

    it('saga: snapshot: fetchById with param', async () => {
        const payload = (await expectSaga(fetchById, 1).run()).toJSON();
        expect(payload).toMatchSnapshot();
    });
});
