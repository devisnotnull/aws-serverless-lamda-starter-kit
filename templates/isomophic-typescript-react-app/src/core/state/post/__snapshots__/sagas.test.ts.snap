// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`post saga: fetchAll saga: snapshot: fetchAllSaga with no page and limit params 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "
    query($options: PageQueryOptions!) {
        posts(options: $options) {
            data {
                id
                title
            }
            meta {
                totalCount
            }
        }
    }
",
          Object {
            "options": Object {
              "paginate": Object {
                "limit": 40,
                "page": 1,
              },
            },
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/fetch",
      },
      "type": "CALL",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          Object {
            "next": "@@redux-saga-test-plan/json/function/<anonymous>",
            "return": "@@redux-saga-test-plan/json/function/<anonymous>",
            "throw": "@@redux-saga-test-plan/json/function/<anonymous>",
          },
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
      "type": "CALL",
    },
  ],
  "getContext": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": "client",
      "type": "GET_CONTEXT",
    },
  ],
}
`;

exports[`post saga: fetchAllStart saga: snapshot: fetchAllSaga with no page and limit params 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "
    query($options: PageQueryOptions!) {
        posts(options: $options) {
            data {
                id
                title
            }
            meta {
                totalCount
            }
        }
    }
",
          Object {
            "options": Object {
              "paginate": Object {
                "limit": 40,
                "page": 1,
              },
            },
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/fetch",
      },
      "type": "CALL",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          Object {
            "next": "@@redux-saga-test-plan/json/function/<anonymous>",
            "return": "@@redux-saga-test-plan/json/function/<anonymous>",
            "throw": "@@redux-saga-test-plan/json/function/<anonymous>",
          },
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
      "type": "CALL",
    },
  ],
  "getContext": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": "client",
      "type": "GET_CONTEXT",
    },
  ],
}
`;

exports[`post saga: fetchByIdStart saga: snapshot: fetchAllSaga with no page and limit params 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "
    query($options: PageQueryOptions!) {
        posts(options: $options) {
            data {
                id
                title
            }
            meta {
                totalCount
            }
        }
    }
",
          Object {
            "options": Object {
              "paginate": Object {
                "limit": 40,
                "page": 1,
              },
            },
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/fetch",
      },
      "type": "CALL",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          Object {
            "next": "@@redux-saga-test-plan/json/function/<anonymous>",
            "return": "@@redux-saga-test-plan/json/function/<anonymous>",
            "throw": "@@redux-saga-test-plan/json/function/<anonymous>",
          },
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
      "type": "CALL",
    },
  ],
  "getContext": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": "client",
      "type": "GET_CONTEXT",
    },
  ],
}
`;
